### ISSUE TITLE
Cannot load such file .../ruby-ll-2.1.1/lib/libll 

### ISSUE THREAD

**Issue by KenRud** [YorickPeterse/ruby-ll#15]
I am trying to use the Oga XML/HTML parser but I have run into this issue. I get the same when I attempt to use ruby-ll on its own. Below is a truncated form of the error:

```
~/.gem/ruby/gems/ruby-ll-2.1.1/lib/ll/setup.rb:6:in `require_relative': cannot load such file -- ~/.gem/ruby/gems/ruby-ll-2.1.1/lib/libll (LoadError)
from ~/.gem/ruby/gems/ruby-ll-2.1.1/lib/ll/setup.rb:6:in `<top (required)>'
.......
```

Ruby-LL was originally installed as a dependency when I installed Oga, but I have tried installing manually myself as well using Rubygems.

**Comment by YorickPeterse** [comment]
How exactly did you install ruby-ll and Oga? For ruby-ll to work its native extension must be compiled, which should happen automatically when running `gem install`.

**Comment by KenRud** [comment]
I added Oga as a Gemfile dependency. When I ran my application it failed with the stated error. I tried also installing Oga with `gem install oga` but that did not help. I also tried installing ruby-ll manually with `gem install ruby-ll` but that did not work either. I made sure to completely uninstall Oga and ruby-ll before each reinstall attempt.

I am running 64-bit Fedora 21 with the following Ruby version:
```
$ ruby -v
ruby 2.1.6p336 (2015-04-13 revision 50298) [x86_64-linux]
```
I also made sure that gcc was installed. Are there any other possible build dependencies that I could need for the native extension to be compiled?

**Comment by YorickPeterse** [comment]
What is the output of `gem install ruby-ll --verbose` and `gem env`?

**Comment by KenRud** [comment]
```
$ gem install ruby-ll --verbose
GET https://api.rubygems.org/latest_specs.4.8.gz
302 Moved Temporarily
GET https://s3.amazonaws.com/production.s3.rubygems.org/latest_specs.4.8.gz
200 OK
HEAD https://api.rubygems.org/api/v1/dependencies
200 OK
GET https://api.rubygems.org/api/v1/dependencies?gems=ruby-ll
200 OK
GET https://api.rubygems.org/api/v1/dependencies?gems=ansi
200 OK
GET https://api.rubygems.org/api/v1/dependencies?gems=ast
200 OK
Downloading gem ruby-ll-2.1.1.gem
GET https://api.rubygems.org/gems/ruby-ll-2.1.1.gem
302 Moved Temporarily
GET https://rubygems.global.ssl.fastly.net/gems/ruby-ll-2.1.1.gem
Fetching: ruby-ll-2.1.1.gem (100%)
200 OK
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/.yardopts
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/LICENSE
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/README.md
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/bin/ruby-ll
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/doc/DCO.md
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/doc/css/common.css
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/doc/driver_architecture.md
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/ext/c/driver.c
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/ext/c/driver.h
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/ext/c/driver_config.c
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/ext/c/driver_config.h
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/ext/c/extconf.rb
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/ext/c/khash.h
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/ext/c/kvec.h
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/ext/c/libll.c
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/ext/c/libll.h
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/ext/c/macros.h
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/ext/java/Libll.java
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/ext/java/org/libll/Driver.java
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/ext/java/org/libll/DriverConfig.java
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/lib/ll.rb
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/lib/ll/ast/node.rb
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/lib/ll/bootstrap/parser.rb
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/lib/ll/branch.rb
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/lib/ll/cli.rb
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/lib/ll/code_generator.rb
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/lib/ll/compiled_configuration.rb
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/lib/ll/compiled_grammar.rb
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/lib/ll/configuration_compiler.rb
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/lib/ll/driver.rb
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/lib/ll/driver_config.rb
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/lib/ll/driver_template.erb
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/lib/ll/epsilon.rb
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/lib/ll/erb_context.rb
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/lib/ll/grammar_compiler.rb
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/lib/ll/lexer.rb
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/lib/ll/message.rb
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/lib/ll/operator.rb
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/lib/ll/parser.rb
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/lib/ll/parser_error.rb
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/lib/ll/rule.rb
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/lib/ll/setup.rb
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/lib/ll/source_line.rb
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/lib/ll/terminal.rb
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/lib/ll/token.rb
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/lib/ll/version.rb
/home/kruddick/.gem/ruby/gems/ruby-ll-2.1.1/ruby-ll.gemspec
Building native extensions.  This could take a while...
/usr/bin/ruby extconf.rb
creating Makefile
make "DESTDIR=" clean
rm -f 
rm -f libll.so  *.o  *.bak mkmf.log .*.time
make "DESTDIR="
gcc -I. -I/usr/include -I/usr/include/ruby/backward -I/usr/include -I.   -fPIC -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -mtune=generic -fPIC -m64 -o driver_config.o -c driver_config.c
gcc -I. -I/usr/include -I/usr/include/ruby/backward -I/usr/include -I.   -fPIC -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -mtune=generic -fPIC -m64 -o driver.o -c driver.c
gcc -I. -I/usr/include -I/usr/include/ruby/backward -I/usr/include -I.   -fPIC -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -mtune=generic -fPIC -m64 -o libll.o -c libll.c
rm -f libll.so
gcc -shared -o libll.so driver_config.o driver.o libll.o -L. -L/usr/lib64 -L. -Wl,-z,relro -fstack-protector -rdynamic -Wl,-export-dynamic  -m64  -lruby  -lpthread -ldl -lcrypt -lm   -lc
make "DESTDIR=" install
/usr/bin/mkdir -p . ./.gem.20150517-8209-111dqg2
exit > .RUBYARCHDIR.time
/usr/bin/install -c -m 0755 libll.so ./.gem.20150517-8209-111dqg2

/home/kruddick/bin/ruby-ll
Successfully installed ruby-ll-2.1.1
Parsing documentation for ruby-ll-2.1.1
Parsing sources...
100% [43/43]  lib/ll/version.rb                   
Installing ri documentation for ruby-ll-2.1.1
Done installing documentation for ruby-ll after 1 seconds
1 gem installed
```

```
$ gem env
RubyGems Environment:
  - RUBYGEMS VERSION: 2.2.2
  - RUBY VERSION: 2.1.6 (2015-04-13 patchlevel 336) [x86_64-linux]
  - INSTALLATION DIRECTORY: /home/kruddick/.gem/ruby
  - RUBY EXECUTABLE: /usr/bin/ruby
  - EXECUTABLE DIRECTORY: /home/kruddick/bin
  - SPEC CACHE DIRECTORY: /home/kruddick/.gem/specs
  - RUBYGEMS PLATFORMS:
    - ruby
    - x86_64-linux
  - GEM PATHS:
     - /home/kruddick/.gem/ruby
     - /usr/share/gems
     - /usr/local/share/gems
  - GEM CONFIGURATION:
     - :update_sources => true
     - :verbose => true
     - :backtrace => false
     - :bulk_threshold => 1000
  - REMOTE SOURCES:
     - https://rubygems.org/
  - SHELL PATH:
     - /home/kruddick/.opam/system/bin
     - /usr/lib64/qt-3.3/bin
     - /usr/local/bin
     - /usr/local/sbin
     - /usr/bin
     - /usr/sbin
     - /bin
     - /sbin
     - /home/kruddick/.local/bin
     - /home/kruddick/bin
```

**Comment by YorickPeterse** [comment]
Can you update to the latest RubyGems (2.4.7) and try again? Other than that I can't really think of anything that might explain why the extension is not in place, especially considering it _was_ apparently compiled (judging based on the output).

**Comment by postmodern** [comment]
Also ran into this on Fedora with ruby 2.1.6 and rubygems 2.2.2. `libll.so` gets compiled/installed into `ext/c/` instead of `lib/`.

<pre>
$ gem install oga
Fetching: ruby-ll-2.1.1.gem (100%)
Building native extensions.  This could take a while...
Successfully installed ruby-ll-2.1.1
Fetching: oga-1.0.1.gem (100%)
Building native extensions.  This could take a while...
Successfully installed oga-1.0.1
Parsing documentation for oga-1.0.1
Installing ri documentation for oga-1.0.1
Parsing documentation for ruby-ll-2.1.1
Installing ri documentation for ruby-ll-2.1.1
Done installing documentation for oga, ruby-ll after 4 seconds
2 gems installed
$ irb
&gt;&gt; require 'oga'
LoadError: cannot load such file -- /home/hal/.gem/ruby/gems/ruby-ll-2.1.1/lib/libll
    /home/hal/.gem/ruby/gems/ruby-ll-2.1.1/lib/ll/setup.rb:6:in `require_relative'
$ gem env
RubyGems Environment:
  - RUBYGEMS VERSION: 2.2.2
  - RUBY VERSION: 2.1.6 (2015-04-13 patchlevel 336) [x86_64-linux]
  - INSTALLATION DIRECTORY: /home/hal/.gem/ruby
  - RUBY EXECUTABLE: /usr/bin/ruby
  - EXECUTABLE DIRECTORY: /home/hal/bin
  - SPEC CACHE DIRECTORY: /home/hal/.gem/specs
  - RUBYGEMS PLATFORMS:
    - ruby
    - x86_64-linux
  - GEM PATHS:
     - /home/hal/.gem/ruby
     - /usr/share/gems
     - /usr/local/share/gems
  - GEM CONFIGURATION:
     - :update_sources => true
     - :verbose => true
     - :backtrace => false
     - :bulk_threshold => 1000
     - "gem" => "--no-rdoc --env-shebang"
$ ls /home/hal/.gem/ruby/gems/ruby-ll-2.1.1/lib/
ll  ll.rb
$ ls /home/hal/.gem/ruby/gems/ruby-ll-2.1.1/ext/c
driver.c         driver_config.o  extconf.rb  libll.c  libll.so
driver_config.c  driver.h         khash.h     libll.h  macros.h
driver_config.h  driver.o         kvec.h      libll.o  Makefile
</pre>

**Comment by postmodern** [comment]
On compiled rubies 2.1.6 and 2.2.2, `libll.so` get's correctly installed into `/home/hal/.gem/ruby/$version/extensions/x86_64-linux/$version-static/ruby-ll-2.1.1/libll.so`.

**Comment by YorickPeterse** [comment]
Please try using the latest version of RubyGems, I suspect this is some sort of bug in RubyGems.

**Comment by postmodern** [comment]
I upgraded system's rubygems to 2.4.7, however the problem remains. Must be due to the configuration of Ubuntu/Fedora's rubygems? Is there a way you could workaround this issue?

**Comment by postmodern** [comment]
<pre>
$ gem which libll.so
/home/hal/.gem/ruby/extensions/x86_64-linux/ruby-ll-2.1.1/libll.so
</pre>

Is there some way of querying Gem or RbConfig for that path and loading `libll.so` from there?

**Comment by YorickPeterse** [comment]
Can you try to build from source and change extconf.rb using the following patch:

```diff
diff --git a/ext/c/extconf.rb b/ext/c/extconf.rb
index 17f6ad9..17401b2 100644
--- a/ext/c/extconf.rb
+++ b/ext/c/extconf.rb
@@ -8,4 +8,4 @@ if ENV['DEBUG']
   $CFLAGS << ' -O0 -g'
 end
 
-create_makefile('libll')
+create_makefile('ll/libll')
```

If that doesn't work I'm not sure what the heck Fedora thinks it's doing, but I'm not going to provide distro specific hacks as I'll only end up tearing my hairs out.

**Comment by YorickPeterse** [comment]
@KenRud Could you try the above patch as well? Curious to see if that fixes the problem.

**Comment by postmodern** [comment]
The patch didn't work. `libll.so` is still not copied into `lib/`. ¯\(°_o)/¯

**Comment by YorickPeterse** [comment]
Well, fuck. I see if I can dig up more information on what the heck is going on here.

**Comment by YorickPeterse** [comment]
Seems this might be related to the extension sharing crap recently introduced in RubyGems: https://github.com/rubygems/rubygems/issues/926#issuecomment-44114814.

**Comment by YorickPeterse** [comment]
I suspect Fedora in this case is enabling this sharing stuff, thus breaking the Gem. I'll just bend over and use `require` in combination with updating the load path during development, I don't have the will to spend hours trying to work around this nonsense.

**Comment by YorickPeterse** [comment]
This has been released as ruby-ll 2.1.2. I'll have to update/release Oga itself as well as it will otherwise suffer from the same problem.

**Comment by YorickPeterse** [comment]
Oga has also been updated and released.
